src/capture.rs

```diff
@@ -71,23 +77,59 @@ pub async fn event(
     }))
 }
 
-pub fn process_single_event(_event: &Event) -> Result<ProcessedEvent> {
-    // TODO: Put actual data in here and transform it properly
+pub fn process_single_event(event: &RawEvent, query: &EventQuery) -> Result<ProcessedEvent> {
+    let distinct_id = match &event.distinct_id {
+        Some(id) => id,
+        None => match event.properties.get("distinct_id").map(|v| v.as_str()) {
+            Some(Some(id)) => id,
+            _ => return Err(anyhow!("missing distinct_id")),
+        },
+    };
+
     Ok(ProcessedEvent {
-        uuid: Uuid::new_v4(),
-        distinct_id: Uuid::new_v4().simple().to_string(),
-        ip: String::new(),
+        uuid: event.uuid.unwrap_or_else(uuid_v7),
+        distinct_id: distinct_id.to_string(),
+        ip: query.client_ip.clone().unwrap_or_default(),
         site_url: String::new(),
         data: String::from("hallo I am some data 😊"),
-        now: String::new(),
+        now: query.now.clone().unwrap_or_default(),
         sent_at: String::new(),
-        token: String::from("tokentokentoken"),
+        token: query.token.clone().unwrap_or_default(),
     })
 }
 
+pub fn extract_and_verify_token(events: &[RawEvent]) -> Result<String, String> {
+    let mut request_token: Option<String> = None;
```

Comment:
I'm fairly against the mutability here, but leave it up to you if it's addressed here or in the future

I'm assuming the reason we do not break out of the loop as soon as we find a token is that we wish to verify that all tokens in a batch are the same?

If so, I'd consider

1. Split out the "get a token from an event" into a function/closure/whatever just for that
2. Map over the list of events, so we can have an iterator of just tokens. This could go into a set, and then assert the set length == 1
3. Assert all items in iterator are equal, error if not
4. Assert token is valid
5. Return

I don't mind doing this in a later PR or pairing on it if you'd like! I know the code to get tokens from our ingestion is awkward.